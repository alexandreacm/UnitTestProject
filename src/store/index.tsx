
type ActionType = {
    type: string;
    payload: any
}

const initialState = {
    status: ''
}

export function reducer(state = initialState, action: ActionType) {
    switch (action.type) {
        case 'setStatus':
            return { ...state, status: action.payload };
        default:
            return state
    }
}