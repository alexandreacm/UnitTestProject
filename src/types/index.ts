export interface Post {
    userId: number;
    id: number;
    title: string;
    completed: boolean;
}

export interface IUser {
    id: number;
    name: string;
    username: string;
    email: string;
}