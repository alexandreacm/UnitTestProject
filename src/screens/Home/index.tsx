import React, { useState, useEffect } from "react";
import { View, Text, StyleSheet, Button, Image, ScrollView } from "react-native";

import { NativeStackHeaderProps } from "@react-navigation/native-stack";

import { useSelector, useDispatch } from 'react-redux';
import { selectStatus } from "../../store";

import sample from '../../assets/images/sample.jpeg';
import { IUser } from "../../types";
// import { api } from "../../services/api";

type Props = {
    navigation: NativeStackHeaderProps;
}

export default function Home({ navigation }: Props) {
    const [status, setStatus] = useState<string>('');
    const [posts, setPosts] = useState<Array<IUser>>([]);
    const [load, setLoad] = useState<Boolean>(false);

    const dispatch = useDispatch();
    // const statusStore = useSelector((state: State) => state.status);
    const statusStore = useSelector(selectStatus);

    // useEffect(() => {
    //     async function fetchData() {

    //         try {
    //             setLoad(true);

    //             const response = await api.get<Array<IUser>>('users');
    //             setPosts(response.data);

    //             setLoad(false);

    //         } catch (err) {
    //             console.log(err);
    //         }
    //     }

    //     fetchData();
    // }, []);

    useEffect(() => {
        setTimeout(() => {
            setStatus('timeout is called');
        }, 1000);
    }, []);

    return (
        <View style={styles.container}>

            <ScrollView>

                <View style={{ flex: 1, alignItems: "center", justifyContent: 'center' }}>
                    <Image source={sample} style={{ width: 150, height: 150 }} />
                </View>


                <View style={{ flex: 1 }}>
                    <Button
                        testID="myButton"
                        title="PressMe"
                        onPress={() => setStatus('button pressed')} />

                    <Button
                        testID="myNavigateButton"
                        title="Navigate to Home"
                        onPress={() => navigation.navigate('Details')} />

                    <Button
                        testID="myButtonRedux"
                        title="Run Redux"
                        onPress={() => dispatch({ type: 'setStatus', payload: 'Redux timeout is called' })} />

                    <Text
                        testID="myText"
                        style={{ textAlign: 'center' }}>{status}</Text>

                    <Text
                        testID="myTextRedux"
                        style={{ textAlign: 'center' }}>{statusStore}</Text>

                    <View style={{ marginTop: 10, backgroundColor: '#f3f3f3', padding: 10 }}>
                        {
                            posts.map((user, idx) => {
                                return (
                                    <Text
                                        key={idx}
                                        testID="posts"
                                        style={{ textAlign: 'center' }}>{user?.name}</Text>
                                )
                            })
                        }
                    </View>
                </View>
            </ScrollView>
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#FB45',
        justifyContent: 'center',
        padding: 12
    },
});
