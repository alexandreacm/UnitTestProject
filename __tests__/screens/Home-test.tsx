
// import 'react-native';
import React from 'react';

import renderer, { create, act } from 'react-test-renderer';
import { render, fireEvent, screen } from '@testing-library/react-native';

import Home from '../../src/screens/Home';

import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { reducer } from '../../src/store';

const navigation: any = {
    navigate: jest.fn()
}

const store = createStore(reducer, { status: 'default store' });

const tree = renderer.create(
    <Provider store={store}>
        <Home navigation={navigation} />
    </Provider>
);

describe('HomeScreen', () => {

    test('Should renders correctly component', () => {
        create(
            <Provider store={store}>
                <Home navigation={navigation} />
            </Provider>
        );
    });

    test('Should render snapshot', () => {
        expect(tree).toMatchSnapshot();
    });

    test('Should show the default text PressMe', () => {
        const { queryAllByText } = render(
            <Provider store={store}>
                <Home navigation={navigation} />
            </Provider>
        );

        // screen.debug();
        queryAllByText('PressMe');
    });

    test('Should render the test statusItem', () => {
        const { getByTestId, debug } = render(
            <Provider store={store}>
                <Home navigation={navigation} />
            </Provider>
        );

        const txtResult = getByTestId('myText');
        const button = getByTestId('myButton');

        fireEvent.press(button);

        expect(txtResult).toBeDefined();

        // debug();
    });

    it('button press with react test renderer', () => {
        const myButton = tree.root.findByProps({ testID: 'myButton' }).props;
        act(() => myButton.onPress());

        const text = tree.root.findByProps({ testID: 'myText' }).props;

        expect(text.children).toEqual('button pressed');
    });

    it('Should timeout to be called', () => {
        jest.runAllTimers();

        const myText = tree.root.findByProps({ testID: 'myText' }).props;
        expect(myText.children).toEqual('timeout is called');
    });

    it('Should useEffect is called with @react-testing-library/react-native', () => {
        const { queryByText } = render(
            <Provider store={store}>
                <Home navigation={navigation} />
            </Provider>)

        queryByText('effect is called');
    });

    it('Should navigate to details page', () => {

        const myButton = tree.root.findByProps({ testID: 'myNavigateButton' }).props;
        myButton.onPress();

        expect(navigation.navigate).toBeCalledWith('Details');
    });

    it('Should navigate to details page with @testing-library', () => {

        const { getByTestId } = render(
            <Provider store={store}>
                <Home navigation={navigation} />
            </Provider>
        );

        const navigateButton = getByTestId('myNavigateButton');
        fireEvent.press(navigateButton);

        expect(navigation.navigate).toBeCalledWith('Details');
    });

    it('Should status store properly', () => {
        const { getByTestId } = render(
            <Provider store={store}>
                <Home navigation={navigation} />
            </Provider>
        );

        const button = getByTestId('myButtonRedux');
        fireEvent.press(button);

        expect(store.getState().status).toEqual('Redux timeout is called');
    });

    it('Should status store properly with react-test-renderer', () => {
        const button = tree.root.findByProps({ testID: 'myButtonRedux' }).props;
        button.onPress();

        expect(store.getState().status).toEqual('Redux timeout is called');
    });
});



