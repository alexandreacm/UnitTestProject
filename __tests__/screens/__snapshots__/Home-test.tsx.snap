// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all tests Should render correctly Snapshot test 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FB45",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Image
    source={1}
  />
  <View
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="myButton"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        PressMe
      </Text>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="myNavigateButton"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Navigate to Home
      </Text>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="myButtonRedux"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Run Redux
      </Text>
    </View>
  </View>
  <Text
    style={
      Object {
        "textAlign": "center",
      }
    }
    testID="myText"
  />
  <Text
    style={
      Object {
        "textAlign": "center",
      }
    }
    testID="myTextRedux"
  >
    default store
  </Text>
</View>
`;
